pkg/kubelet/util/pluginwatcher/example_handler.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pluginwatcher
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"golang.org/x/net/context"
+
+	v1beta1 "k8s.io/kubernetes/pkg/kubelet/util/pluginwatcher/example_plugin_apis/v1beta1"
+	v1beta2 "k8s.io/kubernetes/pkg/kubelet/util/pluginwatcher/example_plugin_apis/v1beta2"
+)
+
+type exampleHandler struct {
+	registeredPlugins        map[string]bool
+	mutex                    sync.Mutex
+	chanForHandlerAckTesting chan error // for testing
+}
+
+func NewExampleHandler() *exampleHandler {
+	return &exampleHandler{
+		chanForHandlerAckTesting: make(chan error),
+		registeredPlugins:        make(map[string]bool),
+	}
+}
+
+func (h *exampleHandler) Cleanup() error {
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
+	h.registeredPlugins = make(map[string]bool)
+	return nil
+}
+
+func (h *exampleHandler) Handler(pluginName string, endpoint string, versions []string, sockPath string) (error, chan bool) {
+
+	// check for supported versions
+	if equal := reflect.DeepEqual([]string{"v1beta1", "v1beta2"}, versions); !equal {
+		return fmt.Errorf("not a supported version"), nil
+	}
+
+	// this handler expect non-empty endpoint as a example
+	if len(endpoint) == 0 {
+		return fmt.Errorf("endpoint can not be empty"), nil
+	}
+
+	_, conn, err := dial(sockPath)
+	if err != nil {
+		return err, nil
+	}
+	defer conn.Close()
+
+	// The plugin handler should be able to use any listed service API version.
+	v1beta1Client := v1beta1.NewExampleClient(conn)
+	v1beta2Client := v1beta2.NewExampleClient(conn)
+
+	// Tests v1beta1 GetExampleInfo
+	_, err = v1beta1Client.GetExampleInfo(context.Background(), &v1beta1.ExampleRequest{})
+	if err != nil {
+		return err, nil
+	}
+
+	// Tests v1beta1 GetExampleInfo
+	_, err = v1beta2Client.GetExampleInfo(context.Background(), &v1beta2.ExampleRequest{})
+	if err != nil {
+		return err, nil
+	}
+
+	// handle registered plugin
+	h.mutex.Lock()
+	_, exist := h.registeredPlugins[pluginName]
+	if exist {
+		h.mutex.Unlock()
+		return fmt.Errorf("plugin %s already registered", pluginName), nil
+	}
+	h.registeredPlugins[pluginName] = true
+	h.mutex.Unlock()
+
+	chanForAckOfNotification := make(chan bool)
+	go func() {
+		select {
+		case <-chanForAckOfNotification:
+			close(chanForAckOfNotification)
+			break
```

Comment:
Hi @figo,

Please remove the `break` keyword. In Go, `select` and `switch` statements do not fall through. Also, there's no point in returning early out of the scope of the `select` statement here since it's not in a loop.